relentless/core/variable.py

```diff
@@ -0,0 +1,311 @@
+__all__ = ['Variable','DesignVariable','DependentVariable',
+           'SameAs',
+           'MixingRule','ArithmeticMean','GeometricMean'
+          ]
+
+import abc
+from enum import Enum
+import numpy as np
+
+class Variable(abc.ABC):
+    """Abstract base class for any variable.
+
+    At minimum, a variable must provide the interface to get its value as a property.
+
+    """
+    @property
+    @abc.abstractmethod
+    def value(self):
+        pass
+
+class DesignVariable(Variable):
+    """Designable variable.
+
+    Represents a quantity that optionally takes lower and upper bounds.
+    When the value of the quantity is set, these bounds will be respected and
+    an internal state will track whether the requested quantity was within or
+    outside these bounds. This is useful for performing constrained
+    optimization and for ensuring physical quantities have meaningful values
+    (e.g., lengths should be positive).
+
+    Parameters
+    ----------
+    value : float or int
+        Value of the variable.
+    const : bool
+        If `False`, the variable can be optimized; otherwise, it is treated as
+        a constant in the optimization (defaults to `False`).
+    low : float or None
+        Lower bound for the variable (`None` means no lower bound).
+    high : float or None
+        Upper bound for the variable (`None` means no upper bound).
+
+    Examples
+    --------
+    A variable with a lower bound::
+
+        >>> v = Variable(value=1.0, low=0.0)
```

Comment:
Can you double check for any instances of `Variable` in this class's docs?

```suggestion
        >>> v = DesignVariable(value=1.0, low=0.0)
```